### PR TITLE
fix: restore policy.train() after eval_policy() (#3668)

### LINKED ISSUE #3668: eval_policy() doesn't restore policy.train() after evaluation — silently disables Dropout / freezes BatchNorm for the rest of training
### Ticket Type

🐛 Bug Report (Something isn't working)

### Environment & System Info

```Shell
lerobot version: 0.5.2 (main @ 81948979)
OS: Darwin 25.3.0 (arm64)
Python: 3.13.5
PyTorch: 2.11.0
Triggered via code inspection — bug is not OS- or version-specific.
```

### Description

`eval_policy` in `src/lerobot/scripts/lerobot_eval.py` calls `policy.eval()` before the rollout but never restores the prior mode on return. When called from the training loop (`lerobot_train.py`'s `eval_policy_all → run_one → eval_one → eval_policy` chain), the policy is left in eval mode for every subsequent training step, which silently:

- disables **Dropout** (no further regularization),
- freezes **BatchNorm** running statistics (no EMA updates).

Under DDP only `is_main_process` runs eval (`lerobot_train.py:527`), so the main rank ends up in eval mode while workers stay in train mode — the all-reduced gradients then combine forward passes computed with different dropout masks and different BN behaviour. That's a real DDP-correctness issue, not just a regularization quirk.

**Scope of impact:**

- Every policy with Dropout in its forward path: the default **ACT** (6 Dropout layers @ p=0.1), Diffusion vision backbone, VQ-BeT, Multi-Task DiT, X-VLA, plus all VLA policies that inherit Dropout from their pretrained HF backbone (PI0 / PI0.5 / PI0-FAST, SmolVLA, GR00T-N1.5, EO1, Wall-X).
- Triggers from the **first eval onward** — i.e. ~80% of training under the default config (`steps=100k, eval_freq=20k`), and ~95–99% of training under the LIBERO / RoboCasa / VLABench example commands in `docs/` (`eval_freq=1k–5k`).
- Policies using only LayerNorm/GroupNorm and no Dropout (TDMPC, RTC) are unaffected. Policies using `FrozenBatchNorm2d` (ACT's ResNet backbone) are immune to the BN-stat half; the Dropout half still bites.

**Empirical impact** (deterministic test in the linked PR): on a tiny Dropout+BatchNorm MLP trained for 500 steps with the bug pattern vs the fixed pattern on identical data and seed, held-out val loss is **10–25% worse** with the bug. Real policies with more layers / longer training are expected to see larger gaps.

A fix + regression tests + quantitative impact test is in PR #3667 

### Context & Reproduction

The bug lives in `eval_policy`'s function body, not in any one CLI command — but here's the call chain and a minimal way to observe it.

**Code path:**

- `src/lerobot/scripts/lerobot_eval.py:307` sets `policy.eval()`
- Never restored before the function returns (`return info` near line 478)
- Reached from training via `eval_policy_all → run_one → eval_one → eval_policy`

**Minimal reproducer** (no env / dataset / hardware required):

```python
import torch
import torch.nn as nn
from unittest.mock import MagicMock, patch

from lerobot.policies.pretrained import PreTrainedPolicy
from lerobot.scripts import lerobot_eval

class FakePolicy(nn.Module):
    def __init__(self):
        super().__init__()
        self.dropout = nn.Dropout(0.5)
    def reset(self): pass

PreTrainedPolicy.register(FakePolicy)  # virtual subclass for isinstance

policy = FakePolicy()
policy.train()
assert policy.training is True  # sanity

with patch.object(lerobot_eval, "rollout", return_value={
    "done":    torch.tensor([[True]], dtype=torch.bool),
    "reward":  torch.tensor([[1.0]], dtype=torch.float32),
    "success": torch.tensor([[True]], dtype=torch.bool),
}):
    lerobot_eval.eval_policy(
        env=MagicMock(num_envs=1), policy=policy,
        env_preprocessor=None, env_postprocessor=None,
        preprocessor=None, postprocessor=None,
        n_episodes=1, max_episodes_rendered=0,
        videos_dir=None, return_episode_data=False, start_seed=None,
    )

print("policy.training after eval_policy:", policy.training)
# Bug:   False   <- Dropout disabled, BN frozen for all subsequent training steps
# Fixed: True
```

**Real-world reproduction:** any `lerobot-train` invocation with `cfg.env` set and `eval_freq < steps` exhibits the bug from step `eval_freq` onward. The published example commands in `docs/source/libero.mdx`, `docs/source/robocasa.mdx`, and `docs/source/vlabench.mdx` all trigger it.

### Relevant logs or stack trace

```Shell
N/A
```

### Checklist

- [x] I have searched existing tickets to ensure this isn't a duplicate.
- [x] I am using the latest version of the `main` branch.
- [x] I have verified this is not an environment-specific problem.

### Additional Info / Workarounds

_No response_

### PR DESCRIPTION
**Pull request by redddddyyyyy**
Fixes #3668

`eval_policy()` sets `policy.eval()` before rollouts but doesn't restore training mode on return. After the first eval checkpoint, every subsequent training step runs with Dropout disabled and BatchNorm stats frozen. The issue report measured 10–25% worse val loss on a small reproducer; under DDP it's worse because only the main rank runs eval, so it ends up in eval mode while workers stay in train mode.

Snapshot `policy.training` before `policy.eval()`, wrap the body in `try/finally`, restore the original mode on the way out. Works whether the function returns normally or raises.

Three regression tests in `tests/scripts/test_eval_policy_train_mode.py`:
- train mode is restored after a normal run
- if the caller already set eval mode, it stays in eval mode (no side effect)
- train mode is restored even when `rollout()` raises

```
pytest tests/scripts/test_eval_policy_train_mode.py -v
```

First and third fail on `main`; all three pass with this fix.

### CODE REVIEW

**Comment by imstevenpmwork** [comment]
Hello @redddddyyyyy 

Thanks for contributing! I'm closing this one as I believe this is taken care by https://github.com/huggingface/lerobot/pull/4162

Kind regards!
